File: packages/shared/src/signaling.ts
import { initializeApp, type FirebaseApp } from 'firebase/app';
import { getAuth, signInAnonymously, type Auth } from 'firebase/auth';
import {
  getDatabase,
  ref,
  set,
  update,
  remove,
  push,
  onValue,
  onChildAdded,
  onDisconnect,
  serverTimestamp,
  type Database,
  type Unsubscribe,
} from 'firebase/database';
import { firebaseConfig } from './firebaseConfig.js';
import type {
  DeviceRecord,
  DeviceWithId,
  SessionDescriptionPayload,
  IceCandidatePayload,
} from './types.js';

let app: FirebaseApp | null = null;
let db: Database | null = null;
let auth: Auth | null = null;

/** Call once per process before using anything else in this module. */
export async function initSignaling(): Promise<Database> {
  if (db) return db;
  app = initializeApp(firebaseConfig);
  db = getDatabase(app);
  auth = getAuth(app);
  await signInAnonymously(auth);
  return db;
}

const roomId = () => firebaseConfig.roomId;

// ---------- devices ----------

function deviceRef(deviceId: string) {
  return ref(db!, `rooms/${roomId()}/devices/${deviceId}`);
}

function devicesRef() {
  return ref(db!, `rooms/${roomId()}/devices`);
}

/**
 * Registers this device as online, arranges for it to flip to 'offline'
 * automatically if the process dies (onDisconnect), and keeps lastSeen fresh.
 * Returns a stop() function and the heartbeat interval handle.
 */
export function registerDevice(
  deviceId: string,
  info: Omit<DeviceRecord, 'status' | 'lastSeen'>
): { stop: () => void } {
  const r = deviceRef(deviceId);
  const online: DeviceRecord = { ...info, status: 'online', lastSeen: Date.now() };
  set(r, online);
  onDisconnect(r).update({ status: 'offline', lastSeen: serverTimestamp() });

  const heartbeat = setInterval(() => {
    update(r, { lastSeen: Date.now(), status: 'online' });
  }, 20_000);

  return {
    stop: () => {
      clearInterval(heartbeat);
      update(r, { status: 'offline', lastSeen: Date.now() });
    },
  };
}

export function listenDevices(cb: (devices: DeviceWithId[]) => void): Unsubscribe {
  return onValue(devicesRef(), (snap) => {
    const val = (snap.val() ?? {}) as Record<string, DeviceRecord>;
    const list: DeviceWithId[] = Object.entries(val).map(([id, d]) => ({ id, ...d }));
    cb(list);
  });
}

/**
 * Forgets a device record outright — e.g. one left behind by a reinstalled
 * or decommissioned agent. A device that comes back online afterwards just
 * re-registers itself (registerDevice writes with set(), not update()).
 */
export function forgetDevice(deviceId: string) {
  remove(deviceRef(deviceId));
}

// ---------- calls (one viewing session) ----------

type CallRole = 'caller' | 'callee';

function callsForDeviceRef(deviceId: string) {
  return ref(db!, `rooms/${roomId()}/calls/${deviceId}`);
}

function callRef(deviceId: string, callId: string) {
  return ref(db!, `rooms/${roomId()}/calls/${deviceId}/${callId}`);
}

function candidatesRef(deviceId: string, callId: string, role: CallRole) {
  return ref(db!, `rooms/${roomId()}/calls/${deviceId}/${callId}/candidates/${role}`);
}

/** Client side: start a new viewing session against a target device. */
export function createCall(targetDeviceId: string): string {
  const r = push(callsForDeviceRef(targetDeviceId));
  set(r, { createdAt: Date.now(), status: 'pending' });
  return r.key as string;
}

export function sendOffer(targetDeviceId: string, callId: string, offer: SessionDescriptionPayload) {
  update(callRef(targetDeviceId, callId), { offer });
}

export function sendAnswer(targetDeviceId: string, callId: string, answer: SessionDescriptionPayload) {
  update(callRef(targetDeviceId, callId), { answer, status: 'accepted' });
}

export function sendIceCandidate(
  targetDeviceId: string,
  callId: string,
  role: CallRole,
  candidate: IceCandidatePayload
) {
  push(candidatesRef(targetDeviceId, callId, role), candidate);
}

export function onRemoteIceCandidates(
  targetDeviceId: string,
  callId: string,
  remoteRole: CallRole,
  cb: (candidate: IceCandidatePayload) => void
): Unsubscribe {
  return onChildAdded(candidatesRef(targetDeviceId, callId, remoteRole), (snap) => {
    cb(snap.val() as IceCandidatePayload);
  });
}

/**
 * Agent side (callee): fires once per new incoming call for this device.
 * The call node is created before its offer is written (createCall vs.
 * sendOffer are separate writes), so this can't just watch for the child
 * being added — it has to watch each new call node's value until an offer
 * shows up, which may already be there or may land moments later.
 */
export function onIncomingCall(
  myDeviceId: string,
  cb: (callId: string, offer: SessionDescriptionPayload) => void
): Unsubscribe {
  const handled = new Set<string>();
  const perCallUnsub = new Map<string, Unsubscribe>();

  const addedUnsub = onChildAdded(callsForDeviceRef(myDeviceId), (snap) => {
    const callId = snap.key as string;
    if (perCallUnsub.has(callId)) return;
    const stop = onValue(callRef(myDeviceId, callId), (s) => {
      const val = s.val();
      if (val?.offer && !handled.has(callId)) {
        handled.add(callId);
        cb(callId, val.offer as SessionDescriptionPayload);
      }
    });
    perCallUnsub.set(callId, stop);
  });

  return () => {
    addedUnsub();
    perCallUnsub.forEach((stop) => stop());
  };
}

/** Either side: watch for the other side setting an answer. */
export function onAnswerSet(
  targetDeviceId: string,
  callId: string,
  cb: (answer: SessionDescriptionPayload) => void
): Unsubscribe {
  return onValue(callRef(targetDeviceId, callId), (snap) => {
    const val = snap.val();
    if (val?.answer) cb(val.answer as SessionDescriptionPayload);
  });
}

/** Either side: watch for the call node being deleted (hangup) by the peer. */
export function onCallEnded(targetDeviceId: string, callId: string, cb: () => void): Unsubscribe {
  return onValue(callRef(targetDeviceId, callId), (snap) => {
    if (!snap.exists()) cb();
  });
}

/** Client hangs up by deleting the call node entirely — agent's onCallEnded fires. */
export function endCall(targetDeviceId: string, callId: string) {
  remove(callRef(targetDeviceId, callId));
}

File: packages/android-agent/app/src/main/java/com/mycamerascontroller/agent/Signaling.kt
package com.mycamerascontroller.agent

import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.ChildEventListener

// Kotlin mirror of packages/shared/src/{types,signaling}.ts — same RTDB schema,
// same "room" concept, so agents and clients on any platform interoperate.

// @JvmOverloads is required so a genuine zero-arg constructor exists for
// Firebase's reflection-based deserializer — a Kotlin all-defaults
// constructor alone does not produce one.
data class DeviceRecord @JvmOverloads constructor(
    var name: String = "",
    var platform: String = "android",
    var status: String = "online",
    var lastSeen: Long = 0L,
)

data class SdpPayload @JvmOverloads constructor(
    var type: String = "",
    var sdp: String = "",
)

data class IceCandidatePayload @JvmOverloads constructor(
    var candidate: String = "",
    var sdpMid: String? = null,
    var sdpMLineIndex: Int? = null,
)

data class DeviceSettings @JvmOverloads constructor(
    var width: Int? = null,
    var height: Int? = null,
    var frameRate: Int? = null,
)

object Signaling {
    private val db by lazy { FirebaseDatabase.getInstance() }
    private var signedIn = false

    /** Older than this and an unanswered call node is presumed abandoned. */
    private const val STALE_CALL_MS = 30_000L

    fun init(onReady: () -> Unit) {
        val auth = FirebaseAuth.getInstance()
        if (auth.currentUser != null) {
            signedIn = true
            onReady()
            return
        }
        auth.signInAnonymously().addOnSuccessListener {
            signedIn = true
            onReady()
        }
    }

    private fun roomRef(path: String) = db.getReference("rooms/${RoomConfig.ROOM_ID}/$path")

    fun registerDevice(deviceId: String, name: String, platform: String = "android") {
        val ref = roomRef("devices/$deviceId")
        val record = DeviceRecord(name = name, platform = platform, status = "online", lastSeen = System.currentTimeMillis())
        ref.setValue(record)
        ref.onDisconnect().updateChildren(mapOf("status" to "offline"))
    }

    fun heartbeat(deviceId: String) {
        roomRef("devices/$deviceId").updateChildren(
            mapOf("status" to "online", "lastSeen" to System.currentTimeMillis())
        )
    }

    /** One-time read of this device's current capture preferences, e.g. right before starting capture. */
    fun getDeviceSettings(deviceId: String, cb: (DeviceSettings) -> Unit) {
        roomRef("deviceSettings/$deviceId").get()
            .addOnSuccessListener { snap -> cb(snap.getValue(DeviceSettings::class.java) ?: DeviceSettings()) }
            .addOnFailureListener { cb(DeviceSettings()) }
    }

    fun setOffline(deviceId: String) {
        roomRef("devices/$deviceId").updateChildren(mapOf("status" to "offline"))
    }

    /**
     * Notifies when this device's own record has been forgotten out from
     * under it (deleted by a client while still running, not just an
     * ordinary disconnect). Only fires after the record has actually been
     * observed to exist, so it can't misfire on the initial empty snapshot
     * before registerDevice's write lands.
     */
    fun onDeviceRemoved(deviceId: String, cb: () -> Unit): ValueEventListener {
        val ref = roomRef("devices/$deviceId")
        var seenExisting = false
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists()) {
                    seenExisting = true
                } else if (seenExisting) {
                    cb()
                }
            }
            override fun onCancelled(error: DatabaseError) {}
        }
        ref.addValueEventListener(listener)
        return listener
    }

    fun stopDeviceRemovedListener(deviceId: String, listener: ValueEventListener) {
        roomRef("devices/$deviceId").removeEventListener(listener)
    }

    /**
     * Fires once per new incoming call for this device.
     * The call node is created before its offer is written (createCall vs.
     * sendAnswer/sendOffer are separate writes on the caller side), so this
     * can't just react to the child being added — it has to watch each new
     * call node's value until an offer shows up, which may already be there
     * or may land a moment later.
     */
    fun onIncomingCall(deviceId: String, cb: (callId: String, offer: SdpPayload) -> Unit): () -> Unit {
        val callsRef = roomRef("calls/$deviceId")
        val handled = mutableSetOf<String>()
        val perCallListeners = mutableMapOf<String, ValueEventListener>()

        val addedListener = object : ChildEventListener {
            override fun onChildAdded(snapshot: DataSnapshot, previousChildName: String?) {
                val callId = snapshot.key ?: return
                if (perCallListeners.containsKey(callId)) return
                val callValueListener = object : ValueEventListener {
                    override fun onDataChange(snap: DataSnapshot) {
                        if (handled.contains(callId)) return
                        val offer = snap.child("offer").getValue(SdpPayload::class.java) ?: return
                        // A call node this stale is a leftover from a caller
                        // that's long gone (e.g. the app closed without
                        // hanging up) — answering it would occupy the "one
                        // call at a time" slot forever with nothing on the
                        // other end, silently blocking every real call after.
                        val createdAt = snap.child("createdAt").getValue(Long::class.java) ?: 0L
                        if (System.currentTimeMillis() - createdAt > STALE_CALL_MS) {
                            handled.add(callId)
                            roomRef("calls/$deviceId/$callId").removeValue()
                            return
                        }
                        handled.add(callId)
                        cb(callId, offer)
                    }
                    override fun onCancelled(error: DatabaseError) {}
                }
                perCallListeners[callId] = callValueListener
                roomRef("calls/$deviceId/$callId").addValueEventListener(callValueListener)
            }
            override fun onChildChanged(snapshot: DataSnapshot, previousChildName: String?) {}
            override fun onChildRemoved(snapshot: DataSnapshot) {}
            override fun onChildMoved(snapshot: DataSnapshot, previousChildName: String?) {}
            override fun onCancelled(error: DatabaseError) {}
        }
        callsRef.addChildEventListener(addedListener)

        return {
            callsRef.removeEventListener(addedListener)
            perCallListeners.forEach { (callId, listener) ->
                roomRef("calls/$deviceId/$callId").removeEventListener(listener)
            }
        }
    }

    fun sendAnswer(targetDeviceId: String, callId: String, answer: SdpPayload) {
        roomRef("calls/$targetDeviceId/$callId").updateChildren(
            mapOf("answer" to answer, "status" to "accepted")
        )
    }

    fun sendIceCandidate(targetDeviceId: String, callId: String, role: String, candidate: IceCandidatePayload) {
        roomRef("calls/$targetDeviceId/$callId/candidates/$role").push().setValue(candidate)
    }

    fun onRemoteIceCandidates(
        targetDeviceId: String,
        callId: String,
        remoteRole: String,
        cb: (IceCandidatePayload) -> Unit,
    ): ChildEventListener {
        val ref = roomRef("calls/$targetDeviceId/$callId/candidates/$remoteRole")
        val listener = object : ChildEventListener {
            override fun onChildAdded(snapshot: DataSnapshot, previousChildName: String?) {
                snapshot.getValue(IceCandidatePayload::class.java)?.let(cb)
            }
            override fun onChildChanged(snapshot: DataSnapshot, previousChildName: String?) {}
            override fun onChildRemoved(snapshot: DataSnapshot) {}
            override fun onChildMoved(snapshot: DataSnapshot, previousChildName: String?) {}
            override fun onCancelled(error: DatabaseError) {}
        }
        ref.addChildEventListener(listener)
        return listener
    }

    fun stopIceListener(targetDeviceId: String, callId: String, remoteRole: String, listener: ChildEventListener) {
        roomRef("calls/$targetDeviceId/$callId/candidates/$remoteRole").removeEventListener(listener)
    }

    /** Fires when the client hangs up (deletes the call node). */
    fun onCallEnded(targetDeviceId: String, callId: String, cb: () -> Unit): ValueEventListener {
        val ref = roomRef("calls/$targetDeviceId/$callId")
        val listener = object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (!snapshot.exists()) cb()
            }
            override fun onCancelled(error: DatabaseError) {}
        }
        ref.addValueEventListener(listener)
        return listener
    }

    fun stopCallEndedListener(targetDeviceId: String, callId: String, listener: ValueEventListener) {
        roomRef("calls/$targetDeviceId/$callId").removeEventListener(listener)
    }
}
